src/main/java/software/amazon/smithy/lsp/language/FoldingRangeHandler.java

```diff
@@ -0,0 +1,198 @@
+     *
+     * @param range         The Range object held the start line and end line info for the folding range
+     * @param foldingRanges The list holds folding ranges for current document
+     */
+    private void addFoldingRange(Range range, List<FoldingRange> foldingRanges) {
+        if (!isFoldable(range)) return;
+        foldingRanges.add(new FoldingRange(range.getStart().getLine(), range.getEnd().getLine()));
+    }
+
+    /**
+     * Create range with start and end character index and then create a fold range object with given range and add it
+     * to the foldingRanges list.
+     *
+     * @param startCharacter The index of the start character of the folding range
+     * @param endCharacter   The index of the start character of the folding range
+     * @param foldingRanges  The list holds folding ranges for current document
+     */
+    private void addFoldingRange(int startCharacter, int endCharacter, List<FoldingRange> foldingRanges) {
+        Range range = document.rangeBetween(startCharacter, endCharacter);
+        addFoldingRange(range, foldingRanges);
+    }
+
+    /**
+     * Use the import range in DocumentImport class to create the folding range directly
+     *
+     * @param documentImports The document import object that held the import range
+     * @param foldingRanges   The list holds folding ranges for current document
+     */
+    private void addFoldingRangeForImports(DocumentImports documentImports, List<FoldingRange> foldingRanges) {
+        Range range = documentImports.importsRange();
+        addFoldingRange(range, foldingRanges);
+    }
+
+    /**
+     * Main function to iterate statements and process the folding range based on the statement type
+     *
+     * @return The list holds folding ranges for current document
+     */
+    private List<FoldingRange> generateFoldingRanges() {
+        List<FoldingRange> foldingRanges = new ArrayList<>();
+
+        addFoldingRangeForImports(documentImports, foldingRanges);
+
+        ListIterator<Syntax.Statement> iterator = statements.listIterator();
+
+        while (iterator.hasNext()) {
+            var statement = iterator.next();
+            switch (statement) {
+                case Syntax.Statement.TraitApplication trait ->
+                        processFoldingRangeForTraitApplication(trait, iterator, foldingRanges);
+
+                case Syntax.Statement.Metadata metadata -> processFoldingRangeForNode(metadata.value(), foldingRanges);
+
+                case Syntax.Statement.Block blk -> processFoldingRangeForBlock(blk, foldingRanges);
+
+                case Syntax.Statement.NodeMemberDef nodeMember ->
+                        processFoldingRangeForNode(nodeMember.value(), foldingRanges);
+
+                // For most of the current statement types just create the folding range based on
+                // its own range
+                default -> addFoldingRange(statement.start(), statement.end(), foldingRanges);
```

Comment:
I wonder if in practice this ever ends up actually adding a range, since most statements (except the ones already checked in this switch) are single line